Core: move validation & data structures together (#176)
tendermint/spec: spec/core/data_structures.md

> The Tendermint blockchains consists of a short list of basic data types:

Reviewer: ```suggestion
The Tendermint blockchains consists of a short list of data types:
```
basic data types are typically integer and Boolean...

| AppHash           | slice of bytes (`[]byte`)     | Arbitrary byte array returned by the application after executing and commiting the previous block. It serves as the basis for validating any merkle proofs that comes from the ABCI application and represents the state of the actual application rather than the state of the blockchain itself. The first block's `block.Header.AppHash` is given by `ResponseInitChain.app_hash`. | This hash is determined by the application, Tendermint can not perform validation on it.                                                                                                                              |

> | LastResultHash    | slice of bytes (`[]byte`)     | `LastResultsHash` is the root hash of a Merkle tree built from `ResponseDeliverTx` responses (`Log`,`Info`, `Codespace` and `Events` fields are ignored). The first block has `block.Header.ResultsHash == MerkleRoot(nil)`, i.e. the hash of an empty input, for RFC-6962 conformance.                                                                                               | Must  be of length 32                                                                                                                                                                                                  |

Reviewer: compared to the other fields, the description reads much more technical and detailed, and less insightful at the same time.

## BlockID

The `BlockID` contains two distinct Merkle roots of the block. The `BlockID` includes these two hashes, as well as the number of parts (ie. `len(MakeParts(block))`)

| Name        | Type                        | Description                                                                                                                                                      | Validation                                                         |
|-------------|-----------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------|
| Hash        | slice of bytes (`[]byte`)   | MerkleRoot of all the fields in the header (ie. `MerkleRoot(header)`.                                                                                            | hash must be of length 32                                          |

> | PartsHeader | [PartsHeader](#PartsHeader) | Used for secure gossiping of the block during consensus, is the MerkleRoot of the complete serialized block cut into parts (ie. `MerkleRoot(MakeParts(block))`). | Must adhere to the validation rules of [PartsHeader](#PartsHeader) |

Reviewer: Should this be partsheader or partsetheader?

|------------------|-----------------------------|---------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------|
| BlockIDFlag      | [BlockIDFlag](#blockidflag) | Represents the validators participation in consensus: Either voted for the block that received the majority, voted for another block, voted nil or did not vote | Must be one of the fields in the [BlockIDFlag](#blockidflag) enum |
| ValidatorAddress | [Address](#address)         | Address of the validator                                                                                      | Must be of length 20                                              |
| Timestamp        | [Time](#time)               | This field will vary from `CommitSig` to `CommitSig`. It represents the timestamp of the validator.               | [Time](#time)                                                     |

> | Signature        | [Signature](#signature)     | Signature corresponding to the validators participation in consensus.                                         | The length of the signature must be > 0 and < than  64            |

Reviewer: This point was never spelled out in the old spec. We should make it more clear. Somehow, 
- a signature, 
- a vote
- height, round, blockID

belong together and there is a check to ascertain that. Somehow, I must be able to check that a signature belongs to a certain height, round, blockID. I never saw that discussed anywhere

Author: This addition would be better suited alongside the block. The validation of commitsig just checks the length of the signature.

Author: added in a secondary validation section under the execution section of the block.

| ValidatorAddress | slice of bytes (`[]byte`)       | Address of the validator                                                                    | Length must be equal to 20                                                                           |
| ValidatorIndex   | int32                           | Index at a specific block height that corresponds to the Index of the validator in the set. | must be > 0                                                                                          |
| Signature        | slice of bytes (`[]byte`)       | Signature by the validator if they participated in consensus for the associated bock.       | Length of signature must be > 0 and < 64                                                             |

## SignedMsgType

Signed message type represents a signed messages in consensus.

```proto
enum SignedMsgType {

  SIGNED_MSG_TYPE_UNKNOWN = 0;
  // Votes
  SIGNED_MSG_TYPE_PREVOTE   = 1;
  SIGNED_MSG_TYPE_PRECOMMIT = 2;

  // Proposal
  SIGNED_MSG_TYPE_PROPOSAL = 32;
}
```

> ## CanonicalVote

Reviewer: I don't understand how this is linked to the rest of the document.

Author: Canonical Vote is used for signing (privval). It can be seen as implementation specific, but the rust client has implemented it due to its use in tmkms.

Reviewer: I see. My point is that if I read the document from top to bottom, at this point I have no idea what a Canonical Vote is. At the beginning of the document there is a bullet list with data types, but it doesn't show there. Also it is not mentioned. So the role of this data type is just not clearly stated here, I guess.

- [`EvidenceData`](#evidence_data)
- [`Evidence`](#evidence)

## Block

A block consists of a header, transactions, votes (the commit),
and a list of evidence of malfeasance (ie. signing conflicting votes).

| Name       | Type                           | Description                                                                                                                                                                          | Validation                                                                                       |
|------------|--------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------|
| Header     | [Header](#header)              | Header corresponding to the block. This field contains information used throughout consensus and other areas of the protocol. To find out what it contains, visit [header] (#header) | Must adhere to the validation rules of [header](#header)                                         |

> | Data       | [Data](#data)                  | Data contains a list of transactions. The contents of the transaction is unknown to Tendermint.                                                                                      | This field can be empty or populated, but the data is unknown to Tendermint                      |

Reviewer 2: You sort of mention that data is unknown to Tendermint twice. On another note we could add that validation of txs is reliant on the application i.e `CheckTx()`

Author: This is true, but the block validation does not rely on checkTx. I can add a minor reference

| Evidence   | [EvidenceData](#evidence_data) | Evidence contains a list of infractions committed by validators.                                                                                                                     | Can be empty, but when populated the validations rules from [evidenceData](#evidence_data) apply |

> | LastCommit | [Commit](#commit)              | `LastCommit` includes one vote for every current validator.  All votes must either be for the previous block, nil or absent. All votes must have a valid signature from the corresponding validator.  The sum total of the voting power of the validators that voted must be greater than 2/3 of the total voting power of the complete validator set. The number of votes in a commit is limited to 10000 (see `types.MaxVotesCount`).                                                                                             | Can be empty for the initial height and must adhere to the validation rules of [commit](#commit).  |

Reviewer 2: Doesn't a commit have to be empty for the initial height? Consider changing "Can" to "Must"

| Height            | int64               | Height is the height for this header.                                                                                                                                                                                                                                                                                                                                                 | Must be > 0, >= initialHeight, and == previous Height+1                                                                                                                                                                |

> | Time              | [Time](#time)       |                   The timestamp is equal to the weighted median of honest validators. Read more on time in the [BFT-time section](../consensus/bft-time.md). Note: the timestamp of a vote must be greater by at least one millisecond than that of the block being voted on.                                                                                                                                                                                                                                                                                                                                                                    | Time must be >= previous header timestamp + consensus parameters TimeIotaMs.  The timestamp of the first block must be equal to the genesis time (since there's no votes to compute the median).  |

Reviewer 2: ```
The timestamp is equal to the weighted median of honest validators
```
should be
```
The timestamp is equal to the weighted median of vote timestamps present in the last commit
```

| BlockID    | [BlockID](#blockid)              | The blockID of the corresponding block.                              | Must adhere to the validation rules of [BlockID](#blockid).                                              |
| Signatures | Array of [CommitSig](#commitsig) | Array of commit signatures that correspond to current validator set. | Length of signatures must be > 0 and adhere to the validation of each individual [Commitsig](#commitsig) |

## CommitSig

`CommitSig` represents a signature of a validator, who has voted either for nil,
a particular `BlockID` or was absent. It's a part of the `Commit` and can be used
to reconstruct the vote set given the validator set.

| Name             | Type                        | Description                                                                                                  | Validation                                                        |
|------------------|-----------------------------|---------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------|

> | BlockIDFlag      | [BlockIDFlag](#blockidflag) | Represents the way the validator voted, if the validator missed the voting period an absent flag will be set. | Must be one of the fields in the [BlockIDFlag](#blockidflag) enum |

Reviewer 2: Suggestion:
```
Represents the way the validator voted, if the validator missed the voting period an absent flag will be set.
```
To
```
Represents the validators participation in consensus: Either voted for the block that received the majority, voted for another block, voted nil or did not vote.
```

Author: ```suggestion
| BlockIDFlag      | [BlockIDFlag](#blockidflag) | Represents the validators participation in consensus: Either voted for the block that received the majority, voted for another block, voted nil or did not vote | Must be one of the fields in the [BlockIDFlag](#blockidflag) enum |
```

| ValidatorAddress | [Address](#address)         | Address of the validator                                                                                      | Must be of length 20                                              |
| Timestamp        | [Time](#time)               | This field will vary from `CommitSig` to `CommitSig`. It represents the timestamp of the validator.               | [Time](#time)                                                     |
| Signature        | [Signature](#signature)     | Signature corresponding to the validators participation in consensus.                                         | The length of the signature must be > 0 and < than  64            |

NOTE: `ValidatorAddress` and `Timestamp` fields may be removed in the future
(see [ADR-25](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-025-commit.md)).

## BlockIDFlag

BlockIDFlag represents which BlockID the [signature](#commitsig) is for.

```go
enum BlockIDFlag {
  BLOCK_ID_FLAG_UNKNOWN = 0;

>   BLOCK_ID_FLAG_ABSENT  = 1;

Reviewer 2: ```suggestion
  BLOCK_ID_FLAG_ABSENT  = 1; // signatures for other blocks are also considered absent
```

  BLOCK_ID_FLAG_COMMIT  = 2;
  BLOCK_ID_FLAG_NIL     = 3;
}
```

## Vote

A vote is a signed message from a validator for a particular block.
The vote includes information about the validator signing it. When stored in the blockchain or propagated over the network, votes are encoded in Protobuf.

| Name             | Type                            | Description                                                                                 | Validation                                                                                           |
|------------------|---------------------------------|---------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|

> | Type             | [SignedMsgType](#signedmsgtype) | [SignedMsgType](#signedmsgtype)                                                             | A Vote is valid if its corresponding fields are included in the enum [signedMsgType](#signedmsgtype) |

Reviewer 2: I would suggest changing the description to:

```
Either prevote or precommit
```
and removing the bottom part ... "There are two types of votes"

Reviewer 2: Also do we want to add the data struct of the Proposal

Author: Yes, there are few things missing. I didn't want bloat this PR, is it okay if I do a followup with more changes?

> 

Reviewer: "Evidence must not have expired" does not really belong here. Or it should be explained differently: The first 4 bullet points are independent of time, that is, I can still validate them in two years from now. The last one depends on time, and is not a validating rule, but rather a guarantee of Tendermint at the time when Evidence is passed to the application.

Author: I can reword this to be specific for block inclusion.

- Evidence must not have expired. The height (and thus the time) is taken from the common height.

> 

Reviewer: The issue about being "expired" is the same as above. I am not sure it belongs to a document about data structures.  
Also, all the other rules implicitly assume validation at a full node against a local copy of the blockchain (e.g., "invalid" and "conflicting header" in bullets 1, 2 and 3). This should be made clear.

|--------|-------------------|-------------------|---------------------------------------------------------------------------------|
| Header | [Header](#Header) | [Header](#header) | Header cannot be nil and must adhere to the [Header](#Header) validation criteria |
| Commit | [Commit](#commit) | [Commit](#commit) | Commit cannot be nil and must adhere to the [Commit](#commit) criteria            |

## ValidatorSet

| Name       | Type                             | Description                                        | Validation                                                                                                        |
|------------|----------------------------------|----------------------------------------------------|-------------------------------------------------------------------------------------------------------------------|
| Validators | Array of [validator](#validator) | List of the active validators at a specific height | The list of validators can not be empty or nil and must adhere to the validation rules of [validator](#validator) |

> | Proposer   | Array of [validator](#validator) | The block proposer for the corresponding block           | The proposer cannot be nil and must adhere to the validation rules of  [validator](#validator)                    |

Reviewer: why is proposer an array? shouldn't there be only one?

- Evidence must not have expired. The height (and thus the time) is taken from the common height.

> ## LightBlock

Reviewer 2: Maybe a short note on what it is and why we have it i.e. LightBlock is the core data structure used by light clients.

## Validator

| Name             | Type                      | Description                                                                                       | Validation                      |
|------------------|---------------------------|---------------------------------------------------------------------------------------------------|---------------------------------|
| Address          | [Address](#address)       | Validators Address                                                                                | Length must be of size 20       |
| Pubkey           | slice of bytes (`[]byte`) | Validators Public Key                                                                             | must be a length greater than 0 |
| VotingPower      | int64                     | Validators voting power                                                                           | cannot be < 0                   |
| ProposerPriority | int64                     | Validators proposer priority. This is used to gauge when a validator is up next to propose blocks |  No validation, value can be negative and positive                               |

## Address

> Address is a type alias of a slice of bytes. The address is calculated by hashing the bytes using sha256 and truncating it to only use the first 20 bytes of the slice.

Reviewer 2: >The address is calculated by hashing the bytes

Hashing the bytes of what? The pub key?
